// @vtex/api/lib/service/worker/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const http_1 = require("http");
const koa_1 = __importDefault(require("koa"));
const koa_compress_1 = __importDefault(require("koa-compress"));
const koa_router_1 = __importDefault(require("koa-router"));
const ramda_1 = require("ramda");
const constants_1 = require("../../constants");
const MetricsAccumulator_1 = require("../../metrics/MetricsAccumulator");
const loaders_1 = require("../loaders");
const console_1 = require("../logger/console");
const logger_1 = require("../logger/logger");
const TracerSingleton_1 = require("../tracing/TracerSingleton");
const tracingMiddlewares_1 = require("../tracing/tracingMiddlewares");
const listeners_1 = require("./listeners");
const handlers_1 = require("./runtime/builtIn/handlers");
const middlewares_1 = require("./runtime/builtIn/middlewares");
const events_1 = require("./runtime/events");
const router_1 = require("./runtime/events/router");
const graphql_1 = require("./runtime/graphql");
const http_2 = require("./runtime/http");
const router_2 = require("./runtime/http/router");
const routes_1 = require("./runtime/http/routes");
const statusTrack_1 = require("./runtime/statusTrack");
const upSignal = () => {
    const data = JSON.stringify({ statusTrack: true });
    const options = {
        headers: {
            'Content-Type': 'application/json',
        },
        host: 'localhost',
        method: 'POST',
        path: '/_up',
        port: 5000,
    };
    const req = http_1.request(options);
    req.write(data);
    req.end();
};
const isUpSignal = (message) => message === constants_1.UP_SIGNAL;
const onMessage = (service) => (message) => {
    if (isUpSignal(message)) {
        upSignal();
        routes_1.logAvailableRoutes(service);
    }
    else if (statusTrack_1.isStatusTrack(message)) {
        statusTrack_1.trackStatus();
    }
    else {
        listeners_1.logger.warn({
            content: message,
            message: 'Master sent message',
            pid: process.pid,
        });
    }
};
const createAppHttpHandlers = ({ config: { routes, clients } }, serviceJSON) => {
    if (routes && clients) {
        return Object.keys(routes).reduce((acc, routeId) => {
            var _a;
            const serviceRoute = (_a = serviceJSON.routes) === null || _a === void 0 ? void 0 : _a[routeId];
            if (!serviceRoute) {
                throw new Error(`Could not find route: ${routeId}. Please add ${routeId} route in your service.json file`);
            }
            const { path: servicePath, public: publicRoute = false, extensible = false, } = serviceRoute;
            if (publicRoute || extensible) {
                acc.pub[routeId] = {
                    handler: http_2.createPublicHttpRoute(clients, routes[routeId], serviceRoute, routeId),
                    path: servicePath,
                };
            }
            else {
                acc.pvt[routeId] = {
                    handler: http_2.createPrivateHttpRoute(clients, routes[routeId], serviceRoute, routeId),
                    path: `/:account/:workspace${servicePath.replace(/\*([^/]*)/g, ':$1*')}`,
                };
            }
            return acc;
        }, {
            pub: {},
            pvt: {},
        });
    }
    return null;
};
const routerFromPrivateHttpHandlers = (routes) => Object.values(routes).reduce((router, { handler, path }) => router.all(path, handler), new koa_router_1.default());
const createAppGraphQLHandler = ({ config: { graphql, clients } }, { routes }) => {
    const route = routes === null || routes === void 0 ? void 0 : routes[graphql_1.GRAPHQL_ROUTE];
    if (graphql && route && clients) {
        return {
            pvt: {
                [graphql_1.GRAPHQL_ROUTE]: {
                    handler: graphql_1.createGraphQLRoute(graphql, clients, route, graphql_1.GRAPHQL_ROUTE),
                    path: `/:account/:workspace${route.path}`,
                },
            },
        };
    }
    return null;
};
const createAppEventHandlers = ({ config: { events, clients } }, serviceJSON) => {
    if (events && clients) {
        return Object.keys(events).reduce((acc, eventId) => {
            var _a;
            const serviceEvent = (_a = serviceJSON.events) === null || _a === void 0 ? void 0 : _a[eventId];
            acc[eventId] = events_1.createEventHandler(clients, eventId, events[eventId], serviceEvent);
            return acc;
        }, {});
    }
    return null;
};
const createRuntimeHttpHandlers = (appEventHandlers, serviceJSON) => ({
    pvt: {
        __events: {
            handler: router_1.routerFromEventHandlers(appEventHandlers),
            path: '/:account/:workspace/_events',
        },
        __healthCheck: {
            handler: handlers_1.healthcheckHandler(serviceJSON),
            path: '/healthcheck',
        },
        __metricsLogger: {
            handler: handlers_1.metricsLoggerHandler,
            path: '/_metrics',
        },
        __statusTrack: {
            handler: statusTrack_1.statusTrackHandler,
            path: '/_status',
        },
        __whoami: {
            handler: handlers_1.whoAmIHandler(serviceJSON),
            path: '/:account/:workspace/_whoami',
        },
    },
});
const scaleClientCaches = (scaleFactor, options) => Object.entries(options || {}).forEach(([name, opts]) => {
    if (opts && opts.memoryCache && scaleFactor > 1) {
        const previous = opts.memoryCache.storage.max;
        const current = previous / scaleFactor;
        opts.memoryCache.storage.max = current;
        console_1.logOnceToDevConsole(`Scaling ${name} cache capacity from ${previous} to ${current}`, logger_1.LogLevel.Warn);
    }
});
exports.startWorker = (serviceJSON) => {
    listeners_1.addProcessListeners();
    const tracer = TracerSingleton_1.TracerSingleton.getTracer();
    const app = new koa_1.default();
    app.proxy = true;
    app
        .use(tracingMiddlewares_1.addTracingMiddleware(tracer))
        .use(middlewares_1.prometheusLoggerMiddleware())
        .use(middlewares_1.addMetricsLoggerMiddleware())
        .use(koa_compress_1.default())
        .use(middlewares_1.recorderMiddleware);
    const service = loaders_1.getService();
    const { config: { clients } } = service;
    if (clients) {
        scaleClientCaches(serviceJSON.workers, clients.options);
    }
    const appHttpHandlers = createAppHttpHandlers(service, serviceJSON);
    const appGraphQLHandlers = createAppGraphQLHandler(service, serviceJSON);
    const appEventHandlers = createAppEventHandlers(service, serviceJSON);
    const runtimeHttpHandlers = createRuntimeHttpHandlers(appEventHandlers, serviceJSON);
    const httpHandlers = [
        appHttpHandlers,
        appGraphQLHandlers,
        runtimeHttpHandlers,
    ]
        .filter(x => x != null)
        .reduce(ramda_1.mergeDeepRight);
    if (httpHandlers === null || httpHandlers === void 0 ? void 0 : httpHandlers.pub) {
        const publicHandlersRouter = router_2.routerFromPublicHttpHandlers(httpHandlers.pub);
        app.use(publicHandlersRouter);
    }
    if (httpHandlers === null || httpHandlers === void 0 ? void 0 : httpHandlers.pvt) {
        const privateHandlersRouter = routerFromPrivateHttpHandlers(httpHandlers.pvt);
        app.use(privateHandlersRouter.routes());
    }
    process.on('message', onMessage(serviceJSON));
    return app;
};
global.metrics = new MetricsAccumulator_1.MetricsAccumulator();
