// @vtex/api/lib/clients/janus/LicenseManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const qs_1 = require("qs");
const JanusClient_1 = require("./JanusClient");
const TWO_MINUTES_S = 2 * 60;
const BASE_URL = '/api/license-manager';
const routes = {
    accountData: `${BASE_URL}/account`,
    resourceAccess: `${BASE_URL}/resources`,
    topbarData: `${BASE_URL}/site/pvt/newtopbar`,
};
const inflightKey = ({ baseURL, url, params }) => {
    return baseURL + url + qs_1.stringify(params, { arrayFormat: 'repeat', addQueryPrefix: true });
};
class LicenseManager extends JanusClient_1.JanusClient {
    getAccountData(VtexIdclientAutCookie) {
        return this.http.get(routes.accountData, {
            forceMaxAge: TWO_MINUTES_S,
            headers: {
                VtexIdclientAutCookie,
            },
            inflightKey,
            metric: 'lm-account-data',
        });
    }
    getTopbarData(VtexIdclientAutCookie) {
        return this.http.get(routes.topbarData, {
            headers: {
                VtexIdclientAutCookie,
            },
            metric: 'lm-topbar-data',
        });
    }
    canAccessResource(VtexIdclientAutCookie, resourceKey) {
        return this.http.get(`${routes.resourceAccess}/${resourceKey}/access`, {
            headers: {
                VtexIdclientAutCookie,
            },
            metric: 'lm-resource-access',
        }).then(() => true, () => false);
    }
}
exports.LicenseManager = LicenseManager;
