// @vtex/node-error-report/build/errorParsing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const http_1 = require("http");
function isRequestInfo(el) {
    return el === null || el === void 0 ? void 0 : el.requestConfig;
}
exports.isRequestInfo = isRequestInfo;
function isInfraErrorData(el) {
    return (el === null || el === void 0 ? void 0 : el.code) && (el === null || el === void 0 ? void 0 : el.source) && (el === null || el === void 0 ? void 0 : el.requestId);
}
exports.isInfraErrorData = isInfraErrorData;
function parseAxiosError(err) {
    var _a;
    const { url, method, headers: requestHeaders, params, data: requestData, timeout: requestTimeout } = err.config;
    const { status, statusText, headers: responseHeaders, data: responseData } = (_a = err.response) !== null && _a !== void 0 ? _a : {};
    return {
        requestConfig: {
            data: requestData,
            headers: requestHeaders,
            method,
            params,
            timeout: requestTimeout,
            url,
        },
        response: err.response
            ? {
                ...(responseData instanceof http_1.IncomingMessage ? { data: '[IncomingMessage]' } : { data: responseData }),
                headers: responseHeaders,
                status,
                statusText,
            }
            : undefined,
    };
}
function parseError(err) {
    if (err.isAxiosError || err.config) {
        return parseAxiosError(err);
    }
    if (typeof err.getDetailsObject === 'function') {
        return err.getDetailsObject();
    }
    return null;
}
exports.parseError = parseError;
