// @vtex/api/lib/clients/external/ID.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ExternalClient_1 = require("./ExternalClient");
const routes = {
    SEND: '/accesskey/send',
    START: '/start',
    VALIDATE: '/accesskey/validate',
    VALIDATE_CLASSIC: '/classic/validate',
};
const VTEXID_ENDPOINTS = {
    STABLE: 'https://vtexid.vtex.com.br/api/vtexid/pub/authentication',
};
const endpoint = (env) => {
    return VTEXID_ENDPOINTS[env] || env;
};
class ID extends ExternalClient_1.ExternalClient {
    constructor(context, opts) {
        super(endpoint(VTEXID_ENDPOINTS.STABLE), context, opts);
        this.getTemporaryToken = () => {
            const metric = 'vtexid-temp-token';
            return this.http.get(routes.START, { metric }).then(({ authenticationToken }) => authenticationToken);
        };
        this.sendCodeToEmail = (token, email) => {
            const params = { authenticationToken: token, email };
            const metric = 'vtexid-send-code';
            return this.http.get(routes.SEND, { params, metric });
        };
        this.getEmailCodeAuthenticationToken = (token, email, code) => {
            const params = {
                accesskey: code,
                authenticationToken: token,
                login: email,
            };
            const metric = 'vtexid-email-token';
            return this.http.get(routes.VALIDATE, { params, metric });
        };
        this.getPasswordAuthenticationToken = (token, email, password) => {
            const params = {
                authenticationToken: token,
                login: email,
                password,
            };
            const metric = 'vtexid-pass-token';
            return this.http.get(routes.VALIDATE_CLASSIC, { params, metric });
        };
    }
}
exports.ID = ID;
