// @vtex/api/lib/service/worker/runtime/events/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tracingMiddlewares_1 = require("../../../tracing/tracingMiddlewares");
const clients_1 = require("../http/middlewares/clients");
const error_1 = require("../http/middlewares/error");
const settings_1 = require("../http/middlewares/settings");
const timings_1 = require("../http/middlewares/timings");
const compose_1 = require("../utils/compose");
const toArray_1 = require("../utils/toArray");
const body_1 = require("./middlewares/body");
const context_1 = require("./middlewares/context");
exports.createEventHandler = (clientsConfig, eventId, handler, serviceEvent) => {
    const { implementation, options } = clientsConfig;
    const middlewares = toArray_1.toArray(handler);
    const pipeline = [
        tracingMiddlewares_1.nameSpanOperationMiddleware('event-handler', eventId),
        context_1.eventContextMiddleware,
        body_1.parseBodyMiddleware,
        clients_1.clients(implementation, options),
        ...((serviceEvent === null || serviceEvent === void 0 ? void 0 : serviceEvent.settingsType) === 'workspace' || (serviceEvent === null || serviceEvent === void 0 ? void 0 : serviceEvent.settingsType) === 'userAndWorkspace' ? [settings_1.getServiceSettings()] : []),
        timings_1.timings,
        error_1.error,
        tracingMiddlewares_1.traceUserLandRemainingPipelineMiddleware(`user-event-handler:${eventId}`),
        contextAdapter(middlewares),
    ];
    return compose_1.compose(pipeline);
};
function contextAdapter(middlewares) {
    return async function middlewareCascade(ctx) {
        const ctxEvent = {
            body: ctx.state.body,
            clients: ctx.clients,
            key: ctx.vtex.eventInfo ? ctx.vtex.eventInfo.key : '',
            metrics: ctx.metrics,
            sender: ctx.vtex.eventInfo ? ctx.vtex.eventInfo.sender : '',
            state: ctx.state,
            subject: ctx.vtex.eventInfo ? ctx.vtex.eventInfo.subject : '',
            timings: ctx.timings,
            vtex: ctx.vtex,
        };
        await compose_1.composeForEvents(middlewares)(ctxEvent);
        ctx.status = 204;
    };
}
