// @vtex/api/lib/tracing/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const opentracing_1 = require("opentracing");
exports.FORMAT_HTTP_HEADERS = opentracing_1.FORMAT_HTTP_HEADERS;
exports.Span = opentracing_1.Span;
var node_error_report_1 = require("@vtex/node-error-report");
exports.ErrorKinds = node_error_report_1.ErrorKindsBase;
var ErrorReport_1 = require("./errorReporting/ErrorReport");
exports.ErrorReport = ErrorReport_1.ErrorReport;
var createSpanReference_1 = require("./spanReference/createSpanReference");
exports.createSpanReference = createSpanReference_1.createSpanReference;
var SpanReferenceTypes_1 = require("./spanReference/SpanReferenceTypes");
exports.SpanReferenceTypes = SpanReferenceTypes_1.SpanReferenceTypes;
var Tags_1 = require("./Tags");
exports.TracingTags = Tags_1.USERLAND_TAGS;
var UserLandTracer_1 = require("./UserLandTracer");
exports.createTracingContextFromCarrier = UserLandTracer_1.createTracingContextFromCarrier;
var utils_1 = require("./utils");
exports.getTraceInfo = utils_1.getTraceInfo;
