// @vtex/api/lib/clients/infra/Router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const InfraClient_1 = require("./InfraClient");
const routes = {
    AvailableIoVersions: '/_io',
    AvailableServiceVersions: (service) => `${routes.AvailableServices}/${service}`,
    AvailableServices: '/_services',
    InstalledIoVersion: (account, workspace) => `/${account}/${workspace}/io`,
    InstalledService: (account, workspace, name) => `/${routes.InstalledServices(account, workspace)}/services/${name}`,
    InstalledServices: (account, workspace) => `/${account}/${workspace}/services`,
};
class Router extends InfraClient_1.InfraClient {
    constructor(ioContext, opts) {
        super('router', ioContext, opts, true);
        this.listAvailableIoVersions = () => {
            return this.http.get(routes.AvailableIoVersions);
        };
        this.getInstalledIoVersion = () => {
            if (!this.context.account || !this.context.workspace) {
                throw new Error('Missing client parameters: {account, workspace}');
            }
            return this.http.get(routes.InstalledIoVersion(this.context.account, this.context.workspace));
        };
        this.installIo = (version) => {
            if (!this.context.account || !this.context.workspace) {
                throw new Error('Missing client parameters: {account, workspace}');
            }
            return this.http.put(routes.InstalledIoVersion(this.context.account, this.context.workspace), { version });
        };
        this.listAvailableServices = () => {
            return this.http.get(routes.AvailableServices);
        };
        this.getAvailableVersions = (name) => {
            return this.http.get(routes.AvailableServiceVersions(name));
        };
        this.listInstalledServices = () => {
            if (!this.context.account || !this.context.workspace) {
                throw new Error('Missing client parameters: {account, workspace}');
            }
            return this.http.get(routes.InstalledServices(this.context.account, this.context.workspace));
        };
        this.installService = (name, version) => {
            if (!this.context.account || !this.context.workspace) {
                throw new Error('Missing client parameters: {account, workspace}');
            }
            return this.http.post(routes.InstalledServices(this.context.account, this.context.workspace), { name, version });
        };
    }
}
exports.Router = Router;
