// @vtex/api/lib/clients/infra/VBase.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const mime_types_1 = __importDefault(require("mime-types"));
const path_1 = require("path");
const zlib_1 = require("zlib");
const HttpClient_1 = require("../../HttpClient");
const InfraClient_1 = require("./InfraClient");
const appId = process.env.VTEX_APP_ID;
const [runningAppName] = appId ? appId.split('@') : [''];
const routes = {
    Bucket: (bucket) => `/buckets/${runningAppName}/${bucket}`,
    Conflicts: (bucket) => `/buckets/${runningAppName}/${bucket}/conflicts`,
    File: (bucket, path) => `${routes.Bucket(bucket)}/files/${path}`,
    Files: (bucket) => `${routes.Bucket(bucket)}/files`,
};
const isVBaseOptions = (opts) => {
    return typeof opts !== 'string' && !(opts instanceof String);
};
class VBase extends InfraClient_1.InfraClient {
    constructor(context, options) {
        super('vbase@2.x', context, options);
        this.getBucket = (bucket) => {
            const inflightKey = HttpClient_1.inflightURL;
            const metric = 'vbase-get-bucket';
            return this.http.get(routes.Bucket(bucket), { metric, inflightKey });
        };
        this.resetBucket = (bucket) => {
            return this.http.delete(routes.Files(bucket), { metric: 'vbase-reset-bucket' });
        };
        this.listFiles = (bucket, opts) => {
            let params = {};
            if (isVBaseOptions(opts)) {
                params = opts;
            }
            else if (opts) {
                params = { prefix: opts };
            }
            const metric = 'vbase-list';
            const inflightKey = HttpClient_1.inflightUrlWithQuery;
            return this.http.get(routes.Files(bucket), { params, metric, inflightKey });
        };
        this.getFile = (bucket, path) => {
            const inflightKey = HttpClient_1.inflightURL;
            const metric = 'vbase-get-file';
            return this.http.getBuffer(routes.File(bucket, path), { metric, inflightKey });
        };
        this.getJSON = (bucket, path, nullIfNotFound, conflictsResolver) => {
            const headers = conflictsResolver ? { 'X-Vtex-Detect-Conflicts': true } : {};
            const inflightKey = HttpClient_1.inflightURL;
            const metric = 'vbase-get-json';
            return this.http.get(routes.File(bucket, path), { nullIfNotFound, metric, inflightKey, headers })
                .catch((error) => {
                const { response } = error;
                if (response && response.status === 409 && conflictsResolver) {
                    return conflictsResolver.resolve();
                }
                throw error;
            });
        };
        this.getFileStream = (bucket, path) => {
            return this.http.getStream(routes.File(bucket, path), { metric: 'vbase-get-file-s' });
        };
        this.saveFile = (bucket, path, stream, gzip = true, ttl) => {
            return this.saveContent(bucket, path, stream, { gzip, ttl });
        };
        this.getFileMetadata = (bucket, path) => {
            return this.http.head(routes.File(bucket, path), { metric: 'vbase-get-file-metadata' });
        };
        this.saveJSON = (bucket, path, data) => {
            const headers = { 'Content-Type': 'application/json' };
            const metric = 'vbase-save-json';
            return this.http.put(routes.File(bucket, path), data, { headers, metric });
        };
        this.saveZippedContent = (bucket, path, stream) => {
            return this.saveContent(bucket, path, stream, { unzip: true });
        };
        this.deleteFile = (bucket, path) => {
            return this.http.delete(routes.File(bucket, path), { metric: 'vbase-delete-file' });
        };
        this.getConflicts = (bucket) => {
            return this.http.get(routes.Conflicts(bucket), { metric: 'vbase-get-conflicts' });
        };
        this.resolveConflict = (bucket, path, content) => {
            const data = [{
                    op: 'replace',
                    path,
                    value: content,
                }];
            return this.http.patch(routes.Conflicts(bucket), data, { metric: 'vbase-resolve-conflicts' });
        };
        this.saveContent = (bucket, path, stream, opts = {}) => {
            if (!stream.pipe || !stream.on) {
                throw new Error(`Argument stream must be a readable stream`);
            }
            const params = opts.unzip ? { unzip: opts.unzip } : {};
            const headers = {};
            let finalStream = stream;
            headers['Content-Type'] = mime_types_1.default.contentType(path_1.basename(path)) || 'application/octet-stream';
            if (opts.gzip) {
                headers['Content-Encoding'] = 'gzip';
                finalStream = stream.pipe(zlib_1.createGzip());
            }
            if (opts.ttl && Number.isInteger(opts.ttl)) {
                headers['X-VTEX-TTL'] = opts.ttl;
            }
            const metric = 'vbase-save-blob';
            return this.http.put(routes.File(bucket, path), finalStream, { headers, params, metric });
        };
        if (runningAppName === '') {
            throw new Error(`Invalid path to access VBase. Variable VTEX_APP_ID is not available.`);
        }
    }
}
exports.VBase = VBase;
