// @vtex/api/lib/clients/infra/Workspaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("../../constants");
const InfraClient_1 = require("./InfraClient");
const routes = {
    Account: (account) => `/${account}`,
    Promote: (account) => `${routes.Workspace(account, constants_1.DEFAULT_WORKSPACE)}/_promote`,
    Workspace: (account, workspace) => `${routes.Account(account)}/${workspace}`,
};
class Workspaces extends InfraClient_1.InfraClient {
    constructor(context, options) {
        super('router', context, options, true);
        this.list = (account) => {
            return this.http.get(routes.Account(account), { metric: 'workspaces-list' });
        };
        this.get = (account, workspace) => {
            return this.http.get(routes.Workspace(account, workspace), { metric: 'workspaces-get' });
        };
        this.set = (account, workspace, metadata) => {
            return this.http.put(routes.Workspace(account, workspace), metadata, { metric: 'workspaces-set' });
        };
        this.create = (account, workspace, production) => {
            return this.http.post(routes.Account(account), { name: workspace, production }, { metric: 'workspaces-create' });
        };
        this.delete = (account, workspace) => {
            return this.http.delete(routes.Workspace(account, workspace), { metric: 'workspaces-delete' });
        };
        this.reset = (account, workspace, metadata = {}) => {
            const params = { reset: true };
            const metric = 'workspaces-reset';
            return this.http.put(routes.Workspace(account, workspace), metadata, { params, metric });
        };
        this.promote = (account, workspace) => {
            return this.http.put(routes.Promote(account), { workspace }, { metric: 'workspaces-promote' });
        };
    }
}
exports.Workspaces = Workspaces;
