// @vtex/api/lib/clients/janus/Segment.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cookie_1 = __importDefault(require("cookie"));
const ramda_1 = require("ramda");
const constants_1 = require("../../constants");
const HttpClient_1 = require("../../HttpClient");
const JanusClient_1 = require("./JanusClient");
const SEGMENT_COOKIE = 'vtex_segment';
const SEGMENT_MAX_AGE_S = 60 * 60; // 60 minutes - segment is actually immutable
const ALLOWED_QUERY_PREFIXES = ['utm', 'cultureInfo', 'supportedLocales', '__bindingId'];
const filterAndSortQuery = (query) => {
    if (!query) {
        return null;
    }
    const filteredKeys = Object.keys(query)
        .filter((k) => !!k && ALLOWED_QUERY_PREFIXES.some((prefix) => k.startsWith(prefix)))
        .sort();
    return filteredKeys.reduce((acc, val) => {
        acc[val] = query[val];
        return acc;
    }, {});
};
const routes = {
    base: '/api/segments',
    segments: (token) => token ? `${routes.base}/${token}` : routes.base,
};
class Segment extends JanusClient_1.JanusClient {
    constructor() {
        super(...arguments);
        /**
         * Get the segment data using the current `ctx.vtex.segmentToken`
         *
         * @memberof Segment
         */
        this.getSegment = (tracingConfig) => this.rawSegment(this.context.segmentToken, undefined, tracingConfig).then(ramda_1.prop('data'));
        /**
         * Get the segment data from this specific segment token
         *
         * @memberof Segment
         */
        this.getSegmentByToken = (token, tracingConfig) => this.rawSegment(token, undefined, tracingConfig).then(ramda_1.prop('data'));
        this.getOrCreateSegment = async (query, token, tracingConfig) => {
            const { data: segmentData, headers: { 'set-cookie': [setCookies], }, } = await this.rawSegment(token, query, tracingConfig);
            const parsedCookie = cookie_1.default.parse(setCookies);
            const segmentToken = ramda_1.prop(SEGMENT_COOKIE, parsedCookie);
            return {
                segmentData,
                segmentToken,
            };
        };
        this.rawSegment = (token, query, tracingConfig) => {
            const { product } = this.context;
            const filteredQuery = filterAndSortQuery(query);
            const metric = token ? 'segment-get-token' : 'segment-get-new';
            return this.http.getRaw(routes.segments(token), ({
                forceMaxAge: SEGMENT_MAX_AGE_S,
                headers: {
                    'Content-Type': 'application/json',
                    [constants_1.PRODUCT_HEADER]: product || '',
                },
                inflightKey: HttpClient_1.inflightUrlWithQuery,
                metric,
                params: {
                    ...filteredQuery,
                    session_path: product || '',
                },
                tracing: {
                    requestSpanNameSuffix: metric,
                    ...tracingConfig === null || tracingConfig === void 0 ? void 0 : tracingConfig.tracing,
                },
            }));
        };
    }
}
exports.Segment = Segment;
